gitops-engine/pkg/cache/cluster.go

```diff
@@ -1063,35 +1215,127 @@ func (c *clusterCache) IterateHierarchyV2(keys []kube.ResourceKey, action func(r
 		}
 		keysPerNamespace[key.Namespace] = append(keysPerNamespace[key.Namespace], key)
 	}
+
+	// Process namespaced resources with standard hierarchy
 	for namespace, namespaceKeys := range keysPerNamespace {
 		nsNodes := c.nsIndex[namespace]
 		graph := buildGraph(nsNodes)
-		visited := make(map[kube.ResourceKey]int)
-		for _, key := range namespaceKeys {
-			visited[key] = 0
+		c.processNamespaceHierarchy(namespaceKeys, nsNodes, graph, visited, action)
```

Comment:
The diff is a little gnarly here, but the contents of this loop now live in c.processNamespaceHierarchy and the logic hasn't changed.